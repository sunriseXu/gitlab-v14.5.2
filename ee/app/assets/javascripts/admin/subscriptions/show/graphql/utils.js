import produce from 'immer';
import { subscriptionHistoryQueries, subscriptionQueries } from '../constants';

export const getLicenseFromData = ({ data } = {}) => data?.gitlabSubscriptionActivate?.license;
export const getErrorsAsData = ({ data } = {}) => data?.gitlabSubscriptionActivate?.errors || [];

export const updateSubscriptionAppCache = (cache, mutation) => {
  const license = getLicenseFromData(mutation);
  if (!license) {
    return;
  }
  const { query } = subscriptionQueries;
  const { query: historyQuery } = subscriptionHistoryQueries;
  const data = produce({}, (draftData) => {
    draftData.currentLicense = license;
  });
  cache.writeQuery({ query, data });
  const subscriptionsList = cache.readQuery({ query: historyQuery });
  const subscriptionListData = produce(subscriptionsList, (draftData) => {
    draftData.licenseHistoryEntries.nodes = [
      license,
      ...subscriptionsList.licenseHistoryEntries.nodes,
    ];
  });
  cache.writeQuery({ query: historyQuery, data: subscriptionListData });
};
