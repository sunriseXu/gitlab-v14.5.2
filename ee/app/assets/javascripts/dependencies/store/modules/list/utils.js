import { REPORT_STATUS } from './constants';

export const hasDependencyList = ({ dependencies }) => Array.isArray(dependencies);

export const hasReportStatus = ({ report }) =>
  Boolean(report && Object.values(REPORT_STATUS).includes(report.status));

export const isValidResponse = ({ data }) =>
  Boolean(data && hasDependencyList(data) && hasReportStatus(data));
