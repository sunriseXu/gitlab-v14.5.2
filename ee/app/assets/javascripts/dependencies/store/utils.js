import listModule from './modules/list';

export const addListType = (store, listType) => {
  const { initialState, namespace } = listType;
  store.registerModule(namespace, listModule());
  store.dispatch('addListType', listType);
  store.dispatch(`${namespace}/setInitialState`, initialState);
};
