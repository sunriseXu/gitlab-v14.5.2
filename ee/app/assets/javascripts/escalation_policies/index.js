import { defaultDataIdFromObject } from 'apollo-cache-inmemory';
import Vue from 'vue';
import VueApollo from 'vue-apollo';
import createDefaultClient from '~/lib/graphql';
import EscalationPoliciesWrapper from './components/escalation_policies_wrapper.vue';

Vue.use(VueApollo);

const apolloProvider = new VueApollo({
  defaultClient: createDefaultClient(
    {},
    {
      cacheConfig: {
        dataIdFromObject: (object) => {
          // eslint-disable-next-line no-underscore-dangle
          if (object.__typename === 'IncidentManagementOncallSchedule') {
            return object.iid;
          }
          return defaultDataIdFromObject(object);
        },
      },
    },
  ),
});

export default () => {
  const el = document.querySelector('.js-escalation-policies');

  if (!el) return null;

  const { emptyEscalationPoliciesSvgPath, projectPath = '' } = el.dataset;

  return new Vue({
    el,
    apolloProvider,
    provide: {
      projectPath,
      emptyEscalationPoliciesSvgPath,
    },
    render(createElement) {
      return createElement(EscalationPoliciesWrapper);
    },
  });
};
