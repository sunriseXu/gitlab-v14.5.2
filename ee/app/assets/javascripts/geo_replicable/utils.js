import createGqClient, { fetchPolicies } from '~/lib/graphql';

/*
  This file uses a NO_CACHE policy due to the need for Geo data to always be fresh.
  The UI this serves is used to watch the "syncing" process of items and their statuses
  will need to be constantly re-queried as the user navigates around to not mistakenly
  think the sync process is broken.
*/

export const gqClient = createGqClient(
  {},
  {
    fetchPolicy: fetchPolicies.NO_CACHE,
  },
);
