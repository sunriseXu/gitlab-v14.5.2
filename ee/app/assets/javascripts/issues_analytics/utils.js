/**
 * This util method takes the global page filters and transforms parameters which
 * are not standardized between the internal issue analytics api and the public
 * issues api.
 *
 * @param {Object} filters the global filters used to fetch issues data
 *
 * @returns {Object} the transformed filters for the public api
 */
export const transformFilters = ({
  label_name: labels = null,
  milestone_title: milestone = null,
  ...restOfFilters
}) => ({
  ...restOfFilters,
  labels,
  milestone,
});
