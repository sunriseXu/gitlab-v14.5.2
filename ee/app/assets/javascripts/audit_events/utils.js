import { parsePikadayDate, pikadayToString } from '~/lib/utils/datetime_utility';
import { AVAILABLE_TOKEN_TYPES, AUDIT_FILTER_CONFIGS, ENTITY_TYPES } from './constants';
import { parseUsername, displayUsername } from './token_utils';

export const getTypeFromEntityType = (entityType) => {
  return AUDIT_FILTER_CONFIGS.find(
    ({ entityType: configEntityType }) => configEntityType === entityType,
  )?.type;
};

export const getEntityTypeFromType = (type) => {
  return AUDIT_FILTER_CONFIGS.find(({ type: configType }) => configType === type)?.entityType;
};

export const parseAuditEventSearchQuery = ({
  created_after: createdAfter,
  created_before: createdBefore,
  entity_type: entityType,
  entity_username: entityUsername,
  author_username: authorUsername,
  ...restOfParams
}) => ({
  ...restOfParams,
  created_after: createdAfter ? parsePikadayDate(createdAfter) : null,
  created_before: createdBefore ? parsePikadayDate(createdBefore) : null,
  entity_type: getTypeFromEntityType(entityType),
  entity_username: displayUsername(entityUsername),
  author_username: displayUsername(authorUsername),
});

export const createAuditEventSearchQuery = ({ filterValue, startDate, endDate, sortBy }) => {
  const entityValue = filterValue.find((value) => AVAILABLE_TOKEN_TYPES.includes(value.type));
  const entityType = getEntityTypeFromType(entityValue?.type);
  const filterData = entityValue?.value.data;

  const params = {
    created_after: startDate ? pikadayToString(startDate) : null,
    created_before: endDate ? pikadayToString(endDate) : null,
    sort: sortBy,
    entity_type: entityType,
    entity_id: null,
    entity_username: null,
    author_username: null,
    // When changing the search parameters, we should be resetting to the first page
    page: null,
  };

  switch (entityType) {
    case ENTITY_TYPES.USER:
      params.entity_username = parseUsername(filterData);
      break;
    case ENTITY_TYPES.AUTHOR:
      params.author_username = parseUsername(filterData);
      break;
    default:
      params.entity_id = filterData;
  }

  return params;
};
