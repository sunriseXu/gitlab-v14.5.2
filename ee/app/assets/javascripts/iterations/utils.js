import { formatDate } from '~/lib/utils/datetime_utility';

const PERIOD_DATE_FORMAT = 'mmm d, yyyy';

/**
 * The arguments are two date strings in formatted in ISO 8601 (YYYY-MM-DD)
 *
 * @returns {string} ex. "Oct 1, 2021 - Oct 10, 2021"
 */
export function getIterationPeriod({ startDate, dueDate }) {
  const start = formatDate(startDate, PERIOD_DATE_FORMAT, true);
  const due = formatDate(dueDate, PERIOD_DATE_FORMAT, true);
  return `${start} - ${due}`;
}
