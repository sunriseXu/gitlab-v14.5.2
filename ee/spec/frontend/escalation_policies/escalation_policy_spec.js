import { GlSprintf, GlIcon, GlCard, GlCollapse } from '@gitlab/ui';
import { shallowMount } from '@vue/test-utils';
import { cloneDeep } from 'lodash';
import { nextTick } from 'vue';
import EditEscalationPolicyModal from 'ee/escalation_policies/components/add_edit_escalation_policy_modal.vue';
import DeleteEscalationPolicyModal from 'ee/escalation_policies/components/delete_escalation_policy_modal.vue';
import EscalationPolicy, { i18n } from 'ee/escalation_policies/components/escalation_policy.vue';

import {
  deleteEscalationPolicyModalId,
  editEscalationPolicyModalId,
} from 'ee/escalation_policies/constants';
import { parsePolicy } from 'ee/escalation_policies/utils';
import mockPolicies from './mocks/mockPolicies.json';

describe('EscalationPolicy', () => {
  let wrapper;
  const escalationPolicy = parsePolicy(cloneDeep(mockPolicies[0]));

  const createComponent = (policy = escalationPolicy) => {
    wrapper = shallowMount(EscalationPolicy, {
      propsData: {
        policy,
        index: 0,
      },
      stubs: {
        GlSprintf,
      },
    });
  };

  beforeEach(() => {
    createComponent();
  });

  afterEach(() => {
    wrapper.destroy();
  });

  const findDeleteModal = () => wrapper.findComponent(DeleteEscalationPolicyModal);
  const findEditModal = () => wrapper.findComponent(EditEscalationPolicyModal);
  const findWarningIcon = () => wrapper.findComponent(GlIcon);
  const findGlCard = () => wrapper.findComponent(GlCard);
  const findGlCollapse = () => wrapper.findComponent(GlCollapse);

  it('renders a policy with rules', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders a policy without rules', () => {
    const policyWithoutRules = {
      id: mockPolicies[0].id,
      name: mockPolicies[0].name,
      description: mockPolicies[0].description,
      rules: [],
    };
    createComponent(policyWithoutRules);
    expect(findWarningIcon().exists()).toBe(true);
    expect(wrapper.text()).toContain(i18n.noRules);
  });

  describe('Modals', () => {
    describe('delete policy modal', () => {
      it('should render a modal and provide it with correct id', () => {
        const modal = findDeleteModal();
        expect(modal.exists()).toBe(true);
        expect(modal.props('modalId')).toBe(
          `${deleteEscalationPolicyModalId}-${escalationPolicy.id}`,
        );
      });
    });

    describe('edit policy modal', () => {
      it('should render a modal and provide it with correct id and isEditMode props', () => {
        const modal = findEditModal();
        expect(modal.exists()).toBe(true);
        expect(modal.props('modalId')).toBe(
          `${editEscalationPolicyModalId}-${escalationPolicy.id}`,
        );
        expect(modal.props('isEditMode')).toBe(true);
      });
    });
  });

  describe('Card collapsing behavior', () => {
    it('adds content body padding when it is expanded', async () => {
      findGlCollapse().vm.$emit('show');
      await nextTick();
      expect(findGlCard().props('bodyClass')).toBe('gl-p-5');
    });

    it('removes content body padding when it is collapsed', async () => {
      findGlCollapse().vm.$emit('hidden');
      await nextTick();
      expect(findGlCard().props('bodyClass')).toBe('gl-p-0');
    });
  });
});
