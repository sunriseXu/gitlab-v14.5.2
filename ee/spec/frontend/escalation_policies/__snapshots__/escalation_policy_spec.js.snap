// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EscalationPolicy renders a policy with rules 1`] = `
<div>
  <gl-card-stub
    bodyclass=""
    class="gl-mt-5"
    footerclass=""
    headerclass="[object Object]"
  >
     
    <gl-collapse-stub
      visible="true"
    >
      <p
        class="gl-text-gray-500 gl-mb-5"
      >
        
        Description 1 lives here
      
      </p>
       
      <div
        class="gl-border-solid gl-border-1 gl-border-gray-100 gl-rounded-base gl-p-5"
      >
        <div
          class="gl-display-flex gl-align-items-center escalation-rule-row gl-mb-5"
        >
          <span
            class="rule-condition gl-md-w-full"
          >
            <gl-icon-stub
              class="gl-mr-3"
              name="clock"
              size="16"
            />
             IF alert is not 
                  acknowledged
                 in 
            <span
              class="gl-font-weight-bold"
            >
              
                    1 mins
                  
            </span>
          </span>
           
          <span
            class="right-arrow gl-display-none gl-lg-display-block gl-flex-shrink-0 gl-mx-5"
            style="width: 4%;"
          >
            <i
              class="right-arrow-head"
            />
          </span>
           
          <span
            class="gl-display-flex gl-align-items-center rule-action"
          >
            <span
              class="gl-text-truncate"
              title=" THEN email on-call user in schedule Schedule to fill in"
            >
              <gl-icon-stub
                class="gl-mr-3"
                name="notifications"
                size="16"
              />
               THEN 
                    email on-call user in schedule
                   
              <span
                class="gl-font-weight-bold"
              >
                
                      Schedule to fill in
                    
              </span>
            </span>
          </span>
        </div>
        <div
          class="gl-display-flex gl-align-items-center escalation-rule-row"
        >
          <span
            class="rule-condition gl-md-w-full"
          >
            <gl-icon-stub
              class="gl-mr-3"
              name="clock"
              size="16"
            />
             IF alert is not 
                  resolved
                 in 
            <span
              class="gl-font-weight-bold"
            >
              
                    2 mins
                  
            </span>
          </span>
           
          <span
            class="right-arrow gl-display-none gl-lg-display-block gl-flex-shrink-0 gl-mx-5"
            style="width: 8%;"
          >
            <i
              class="right-arrow-head"
            />
          </span>
           
          <span
            class="gl-display-flex gl-align-items-center rule-action"
          >
            <span
              class="gl-text-truncate"
              title=" THEN email user Lena"
            >
              <gl-icon-stub
                class="gl-mr-3"
                name="notifications"
                size="16"
              />
               THEN 
                    email user
                   
              <gl-token-stub
                variant="default"
                viewonly="true"
              >
                <gl-avatar-stub
                  alt="avatar"
                  entityid="0"
                  entityname=""
                  shape="circle"
                  size="16"
                  src="avatar.com/lena.png"
                />
                
                      Lena
                    
              </gl-token-stub>
            </span>
          </span>
        </div>
      </div>
    </gl-collapse-stub>
  </gl-card-stub>
   
  <delete-escalation-policy-modal-stub
    escalationpolicy="[object Object]"
    modalid="deleteEscalationPolicyModal-37"
  />
   
  <edit-escalation-policy-modal-stub
    escalationpolicy="[object Object]"
    iseditmode="true"
    modalid="editEscalationPolicyModal-37"
  />
</div>
`;
